### PR TITLE
MFU calculator

### LINKED ISSUE #39: Compute throughput and MFU
Rather than manually calculating the step time, throughput, MFU etc, we should add a library to programmatically do that.

### PR DESCRIPTION
**Pull request by tengyifei**
This change adds MFU calculation functionality. It's not wired up to any trainer in any way because we don't have a standard schema for spelling out the LLM layer shapes etc.

Fixes #39.

### CODE REVIEW

File: torchprime/metrics/mfu.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+  )
+  embedding_flops = (
+    2
+    * config.per_device_batch_size
+    * config.max_target_length
+    * config.emb_dim
+    * config.vocab_size
+  )
+
+  # multiply by 3 for both feed forward and back propagation flops
```
Comment (bhavya01) [inline torchprime/metrics/mfu.py:108]: Multiply by 3 is for gradient checkpointing right?

File: torchprime/metrics/mfu.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+  total_tflops = learnable_weight_tflops + attention_tflops
+
+  if log:
+    print(
+      "Per train step:\n",
+      f"Total TFLOPs: {total_tflops:.2f} \n",
+      f"split as {100 * learnable_weight_tflops/total_tflops:.2f}% learnable weight flops",
+      f"and {100 * attention_tflops/total_tflops:.2f}% attention flops",
+    )
+  return total_tflops, learnable_weight_tflops, attention_tflops
```
Comment (bhavya01) [inline torchprime/metrics/mfu.py:131]: nit: Should we just return total_tflops? or Are the other things returned just for testing?

**Review by bhavya01 — APPROVED** [review]

File: torchprime/metrics/mfu.py (hunk 1 above)
Comment (tengyifei) [reply torchprime/metrics/mfu.py:108]: I asked ChatGPT and I believe that's because the backward pass computes two gradients (wrt weights, wrt activations) whereas the forward pass only computes one output (the activation): https://chatgpt.com/share/6792a528-cab0-800b-9ba7-d45718dfaeb3

Based on pure analysis, I think it shouldn't be the case that gradient checkpointing is accounted for in the MFU calculation. Otherwise, we'll be artificially inflating the MFU just by activating checkpointing.

Commented in the code.

File: torchprime/metrics/mfu.py (hunk 2 above)
Comment (tengyifei) [reply torchprime/metrics/mfu.py:131]: Removed. MaxText did this and I just copied it :)
